conductr_cli/conductr_backup.py

```diff
@@ -0,0 +1,227 @@
+            initial_bundle_core_info = BundleCoreInfo.from_bundles(initial_bundles_json)
+
+            if not bundle_id_or_name:
+
+                for info in initial_bundle_core_info:
+                    backup_bundle(args, backup_directory, info)
+
+                final_bundles_raw = bundles(args)
+                final_bundles_json = json.loads(final_bundles_raw)
+                final_bundle_core_info = BundleCoreInfo.from_bundles(final_bundles_json)
+
+                process_removed_bundles(backup_directory, initial_bundle_core_info, final_bundle_core_info)
+                process_added_bundles(args, backup_directory, initial_bundle_core_info, final_bundle_core_info)
+                backup_bundle_json(backup_directory, final_bundles_raw)
+
+            else:
+                bundle_info = BundleCoreInfo.filter_by_bundle_id(initial_bundle_core_info, bundle_id_or_name)
+
+                if not bundle_info:
+                    raise ConductBackupError('Bundle {} was not found.'.format(bundle_id_or_name))
+
+                filtered_bundle = filter_bundles_by_id_or_name(initial_bundles_json, bundle_id_or_name)
+                backup_bundle_json(backup_directory, json.dumps(filtered_bundle))
+                backup_bundle(args, backup_directory, bundle_info)
+
+            backup_members(args, backup_directory)
+            backup_agents(args, backup_directory)
+
+            compress_backup(args.output_path, backup_directory)
+
+    finally:
+        remove_backup_directory(backup_directory)
+
+
+def compress_backup(output_path, backup_directory):
+    output_file = open(output_path, 'wb') if output_path else sys.stdout.buffer
+    with tempfile.NamedTemporaryFile() as zip_file_data:
+        with zipfile.ZipFile(zip_file_data, 'w') as zip_file:
+            dir_to_zip(backup_directory, zip_file, '.', None)
+        zip_file_data.flush()
+        zip_file_data.seek(0)
+
+        write_with_digest(zip_file_data, output_file)
+    output_file.flush()
+
+
+def process_removed_bundles(backup_path, initial_bundle_core_info, final_bundle_core_info):
+    bundles_removed = BundleCoreInfo.diff(initial_bundle_core_info, final_bundle_core_info)
+    for bundle_info in bundles_removed:
+        remove_bundle(backup_path, bundle_info)
+
+
+def process_added_bundles(args, backup_path, initial_bundle_core_info, final_bundle_core_info):
+    bundles_added = BundleCoreInfo.diff(final_bundle_core_info, initial_bundle_core_info)
+    for bundle_info in bundles_added:
+        backup_bundle(args, backup_path, bundle_info)
+
+
+def backup_bundle(args, backup_path, bundle_core_info: BundleCoreInfo):
+    log = logging.getLogger(__name__)
+    log.debug('Processing bundle : {}'.format(bundle_core_info.bundle_id))
+
+    bundle_files_response = bundle_files(args, bundle_core_info.bundle_id)
+    bundle_validated = backup_bundle_file(backup_path, bundle_files_response[0], bundle_core_info)
+    bundle_conf_validated = True
+
```

Comment:
@fsat previously this was False which would never get set correctly, if the bundle had no config.